### DISCUSSION TITLE
a large dags

### DISCUSSION THREAD

**Discussion by qinyang666** [apache/airflow#14371] [Q&A]
I have 1000000 dags, How should I genarate it to ensure performance, and does the airflow support such a large dags

**Comment by turbaszek** [comment]
This is a huge deployment... but here are some of my thoughts:

1. Consider using Airflow 2.0.1 - it's vastly improves performance and implements HA scheduler which may be crucial in your case.
1. Consider using Kubernetes or Celery executor (or CeleryKubernetesExecutor in Airflow 2.0)  - this helps with scalability. The executor type will depend on type of your tasks.
1. Avoid top level requests (Variable.get, http requests, db queries) in your DAGs to avoid performance degradation during scheduling loop.

Assuming you want to run 1 000 000 DAGs with `T` tasks each at the same moment you will need `(1000000 * T) / C` celery workers with concurrency of `C` - this is huge. According to our survey biggest reported use-case has about 15000 DAGs...

**Reply by qinyang666** [reply]
Actually I only have 100 DAG, but there are 20,000 different parameters that need to be run, and I'm thinking of trigger 20,000 times at a time using the triggerDagrunoOperator. But generating 20,000 TriggerDagrunoOperators in a DAG is not a good idea

**Reply by turbaszek** [reply]
You mean there's 100 DAG and each has to be run for 20 000 different configurations? 

In general if you have a DAG that has to be replicated many times but some configurations differs for each replicas I would suggest creating a DAG template and then generate dag files using jinja or something. It's simple to add new DAGs, update existing ones and you keep configuration in one place - your script.

**Reply by turbaszek** [reply]
@qinyang666 did it answer your question?

**Reply by qinyang666** [reply]
> @qinyang666 did it answer your question?

Jinja template just allows me to use different configurations flexibly. Then I still need to trigger it or use it to generate multiple DAG files. When I triggered the DAG 20000 times, it took me a lot of time to complete all the tasks. It's not even as good as multithreading

**Reply by turbaszek** [reply]
> it took me a lot of time to complete all the tasks

Can you please describe your deployment? From what I understand you have thousands of tasks so proper setup would be crucial.

**Reply by qinyang666** [reply]
> Can you please describe your deployment? From what I understand you have thousands of tasks so proper setup would be crucial.

scheduler 8C * 3 （cpu： about70%）
worker     4C * 4 （cpu： about83%） 
worker_concurrency ：16
parallelism  1000
max_active_runs_per_dag  1000
dag_concurrency  1000

3dag * 1000dagrun  total: 3000dagrun
It takes about four minutes to complete all the tasks

**Reply by turbaszek** [reply]
Well, the time is quite good to be honest (4 minuts for at least 3000 tasks).

@qinyang666 what version of Airflow do you use? If 1.10 then I would suggest migrating to 2.0 which improves the performance drastically. See:
https://medium.com/databand-ai/improving-performance-of-apache-airflow-scheduler-507f4cb6462a
https://www.astronomer.io/blog/airflow-2-scheduler


> as good as multithreading

Multithreading works on single machine, Airflow with Celery is a distributed system so it will never be as instant as spawning new process...
